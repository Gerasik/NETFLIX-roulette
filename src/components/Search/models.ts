import * as SearchModels from 'containers/Search/models';
import * as BodyModels from 'containers/Body/models';

interface SearchData {
  searchString: string;
  searchBy: string;
  sortBy: string;
}

export interface SearchProps {
  searchData: SearchData;
  changeSearchString: SearchModels.ActionCreator.CHANGE_SEARCH_STRING;
  changeSearchBy: SearchModels.ActionCreator.CHANGE_SEARCH_BY;
  setData: BodyModels.ActionCreator.SetData;
}
