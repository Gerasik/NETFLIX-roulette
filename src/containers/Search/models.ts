import Immutable from 'immutable';

import * as Models from 'models';

export type State = Immutable.Record<{
  searchString: string;
  searchBy: string;
  sortBy: string;
}>;

export namespace Payload {
  export type CHANGE_SEARCH_STRING = string;
  export type CHANGE_SEARCH_BY = string;
  export type CHANGE_SORT_BY = string;
}

export namespace Action {
  export type CHANGE_SEARCH_STRING = Models.IAction<Payload.CHANGE_SEARCH_STRING>;
  export type CHANGE_SEARCH_BY = Models.IAction<Payload.CHANGE_SEARCH_BY>;
  export type CHANGE_SORT_BY = Models.IAction<Payload.CHANGE_SORT_BY>;
}
export namespace ActionCreator {
  export type CHANGE_SEARCH_STRING = (
    payload: Payload.CHANGE_SEARCH_STRING
  ) => Action.CHANGE_SEARCH_STRING;
  export type CHANGE_SEARCH_BY = (payload: Payload.CHANGE_SEARCH_BY) => Action.CHANGE_SEARCH_BY;
  export type CHANGE_SORT_BY = (payload: Payload.CHANGE_SORT_BY) => Action.CHANGE_SORT_BY;
}
