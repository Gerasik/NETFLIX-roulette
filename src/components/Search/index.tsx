import React, { FunctionComponent } from 'react';
import classNames from 'classnames';

import { Action } from 'containers/Search/models';
import styles from './style.module.scss';
import { SearchProps } from './models';

const Search: FunctionComponent<SearchProps> = ({
  searchData,
  changeSearchString,
  changeSearchBy,
  setData,
}) => {
  const { searchString, searchBy, sortBy } = searchData;

  function handleClickSearch(): void {
    fetch(
      `https://reactjs-cdp.herokuapp.com/movies?sortBy=${sortBy}&sortOrder=asc&search=${searchString}&searchBy=${searchBy}&limit=6`
    )
      .then(response => response.json())
      .then(response => {
        setData(response);
      });
  }

  return (
    <>
      <p className={styles.label}>find your movie</p>
      <div className={styles['search-form']}>
        <input
          type="text"
          placeholder="Search"
          className={styles['search-text']}
          onChange={(event): Action.CHANGE_SEARCH_STRING => changeSearchString(event.target.value)}
          value={searchString}
        />
        <button type="button" className={styles['search-button']} onClick={handleClickSearch}>
          search
        </button>
      </div>
      <div className={styles['search-by']}>
        <span>search by</span>
        <button
          type="button"
          className={classNames(styles['by-title'], { [styles.active]: searchBy === 'title' })}
          onClick={(): Action.CHANGE_SEARCH_BY => changeSearchBy('title')}
        >
          title
        </button>
        <button
          type="button"
          className={classNames(styles['by-genre'], { [styles.active]: searchBy === 'genres' })}
          onClick={(): SearchModels.Action.CHANGE_SEARCH_BY => changeSearchBy('genres')}
        >
          genre
        </button>
      </div>
    </>
  );
};

export default Search;
