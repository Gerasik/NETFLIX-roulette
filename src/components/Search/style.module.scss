@import '../../styles/colors';

$border-radius-size: 0.3em;

:root {
  color: $font-color;
  width: 100%;
  button {
    border: none;
    color: $font-color;
    text-transform: uppercase;
  }
  .lable {
    font-size: 2em;
    margin: 0.4em 0;
  }
  .lable,
  .search-by {
    font-weight: 100;
    text-transform: uppercase;
  }
  .search-form {
    display: flex;
    justify-content: stretch;
    input,
    button {
      font-size: 1.4em;
      border-radius: $border-radius-size;
    }
    input {
      color: $font-color;
      width: 100%;
      margin-right: 0.5em;
      border: none;
      background-color: rgba($color: $main-color, $alpha: 0.8);
      padding: 0.5em 1em;
    }
    button {
      padding: 0.5em 2em;
      background-color: $main-font-color;
    }
  }
  .search-by {
    font-size: 0.9em;
    margin: 1.3em 0;
    span {
      padding-right: 1em;
    }
    button {
      font-size: 0.9em;
      padding: 0.5em 2em;
      background-color: rgba($color: $footer-color, $alpha: 0.6);
      &.by-title {
        border-radius: $border-radius-size 0em 0em $border-radius-size;
      }
      &.by-gengre {
        border-radius: 0 $border-radius-size $border-radius-size 0;
      }
      &:hover {
        background-color: $main-font-color;
      }
    }
    .active {
      background-color: $main-font-color;
    }
  }
}
