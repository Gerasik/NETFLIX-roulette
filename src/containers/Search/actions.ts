import ActionType from './constants';
import { ActionCreator } from './models';

export const changeSearchString: ActionCreator.CHANGE_SEARCH_STRING = payload => ({
  type: ActionType.CHANGE_SEARCH_STRING,
  payload,
});

export const changeSearchBy: ActionCreator.CHANGE_SEARCH_BY = payload => ({
  type: ActionType.CHANGE_SEARCH_BY,
  payload,
});

export const changeSortBy: ActionCreator.CHANGE_SORT_BY = payload => ({
  type: ActionType.CHANGE_SORT_BY,
  payload,
});
