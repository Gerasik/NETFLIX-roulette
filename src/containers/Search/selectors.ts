import { createSelector } from 'reselect';

import { State } from './models';

const searchInfo = (state): State => {
  return state.get('searchReducer') as State;
};

export const searchSortBy = (state): string => {
  return state.get('searchReducer').get('sortBy');
};

export const searchData = createSelector(searchInfo, searchInfo => ({
  searchString: searchInfo.get('searchString'),
  searchBy: searchInfo.get('searchBy'),
  sortBy: searchInfo.get('sortBy'),
}));
