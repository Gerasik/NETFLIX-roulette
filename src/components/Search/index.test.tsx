import React from 'react';
import { shallow } from 'enzyme';

import Search from '.';

it('render correctly date component', () => {
  const wrapper = shallow(<Search />);
  expect(wrapper).toMatchSnapshot();
});

describe('<Search />', () => {
  it('renders an `footer`', () => {
    const wrapper = shallow(<Search />);
    expect(wrapper.find('input').prop('placeholder')).toBe('Search');
  });
});
