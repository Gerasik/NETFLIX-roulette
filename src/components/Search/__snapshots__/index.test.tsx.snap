// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly date component 1`] = `
<Fragment>
  <p
    className="label"
  >
    find your movie
  </p>
  <div
    className="search-form"
  >
    <input
      className="search-text"
      placeholder="Search"
      type="text"
    />
    <button
      className="search-button"
      type="button"
    >
      search
    </button>
  </div>
  <div
    className="search-by"
  >
    <span>
      search by
    </span>
    <button
      className="by-title active"
      type="button"
    >
      title
    </button>
    <button
      className="by-genre"
      type="button"
    >
      genre
    </button>
  </div>
</Fragment>
`;
